sw/device/lib/base/macros.h

```diff
@@ -162,4 +162,82 @@
+ * program counter value an inline assembly block at this location would
+ * see.
+ *
+ * The primary intention of this macro is to allow for peripheral tests to be
+ * written that want to assert that a specific program counter reported by the
+ * hardware corresponds to actual code that executed.
+ *
+ * For example, suppose that we're waiting for an interrupt to fire, and want
+ * to verify that it fired in a specific region. We have two volatile globals:
+ * `irq_happened`, which the ISR sets before returning, and `irq_pc`, which
+ * the ISR sets to whatever PC the hardware reports it came from. The
+ * code below tests that the reported PC matches expectations.
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kIrqWaitStart);
+ * enable_interrupts();
+ * while(!irq_happened) {
+ *   wait_for_interrupt();
+ * }
+ * OT_ADDRESSABLE_LABEL(kIrqWaitEnd);
+ *
+ * CHECK(irq_pc >= &kIrqWaitStart && irq_pc <= &IrqWaitEnd);
+ * ```
+ *
+ * Note that this only works if all functions called between the two labels
+ * are actually inlined; if the interrupt fires inside of one of those two
+ * functions, it will appear to not have the right PC, even though it is
+ * logically inside the pair of labels.
+ *
+ * # Volatile Semantics
+ *
+ * This has the same semantics as a `volatile` inline assembly block: it
+ * may be reordered with respect to all operations except other volatile
+ * operations (i.e. volatile assembly and volatile read/write, such as
+ * MMIO). For example, in the following code, `kBefore` and `kAfter` can
+ * wind up having the same address:
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kBefore);
+ * x += 5;
+ * OT_ADDRESSABLE_LABEL(kAfter);
+ * ```
+ *
+ * Because it can be reordered and the compiler is free to emit whatever
+ * instructions it likes, comparing a program counter value obtained from
+ * the hardware with a symbol emitted by this macro is brittle (and,
+ * ultimately, futile). Instead, it should be used in pairs to create a
+ * range of addresses that a value can be checked for being within.
+ *
+ * This macro should not be used on the host side. It will compile, but will
+ * probably not provide any meaningful values. In the future, it may start
+ * producing a garbage value on the host side.
+ *
+ * # Symbol Access
+ *
+ * The symbol reference `kName_` will only be scoped to the current block
+ * in a function, but it can be redeclared with
+ *
+ * ```
+ * extern const char kName_[];
+ * ```
+ *
+ * if needed elsewhere (although this should be avoided for readability's sake).
+ * The name of the constant is global to the program, so it should be given a
+ * name that is not going to collide (such as `kMyFunctionBeforeLoop`).
+ *
+ * # Example Uses
+ */
+#define OT_ADDRESSABLE_LABEL(kName_)                                         \
+  extern const char kName_[];                                                \
+  asm volatile(".global " #kName_ "; " #kName_ ":;");                        \
```

Comment:
Shouldn't this be a `.local` directive instead?

Reply:
Neat, I did not know there was a way to specify that. It seems elf-only, so if someone's mac build breaks we'll sort out a workaround for them when the time comes.